sklearn/metrics/_ranking.py

```diff
@@ -1541,15 +1556,21 @@ def _tie_averaged_dcg(y_true, y_score, discount_cumsum):
     European conference on information retrieval (pp. 414-421). Springer,
     Berlin, Heidelberg.
     """
-    _, inv, counts = np.unique(-y_score, return_inverse=True, return_counts=True)
-    ranked = np.zeros(len(counts))
-    np.add.at(ranked, inv, y_true)
-    ranked /= counts
-    groups = np.cumsum(counts) - 1
-    discount_sums = np.empty(len(counts))
+    xp, _, device = get_namespace_and_device(y_true, y_score)
+    # TODO: use unique_all when pytorch supports it
+    # _, _, inv, counts = xp.unique_all(-y_score)
+    _, inv = xp.unique_inverse(-y_score)
+    _, counts = xp.unique_counts(-y_score)
```

Comment:
How much of a performance hit do we get performing 'unique' twice? Would it be worth making a helper that uses `unique_all` unless the xp is torch?